### PR TITLE
Detect ELF-32 at runtime, rather than compile time.

### PR DESCRIPTION
**Pull request by ahmedcharles**
WIP: don't merge.

This is sort of a hacked up runtime implementation of ELF-32 support, primarily so @jmi2k can take a look and see if it works and satisfies their requirements.

I'm planning other changes that will make this much cleaner, but they're not ready yet.

### CODE REVIEW

**Comment by jmi2k** [comment]
I made a quick test and it works flawless. Good job!

File: src/elf_sections.rs [hunk 1]
```diff
@@ -53,22 +57,27 @@ impl StringTable {
...
     type Item = &'static ElfSection;
+
     fn next(&mut self) -> Option<&'static ElfSection> {
         if self.remaining_sections == 0 {
             None
         } else {
-            let section = self.current_section;
+            let section: &'static ElfSection = match self.entry_size {
+                40 => unsafe { &*(self.current_section as *const _ as *const ElfSection32) },
+                64 => unsafe { &*(self.current_section as *const _ as *const ElfSection64) },
```
Comment (4e554c4c) [inline src/elf_sections.rs]: Is the  `as *const _` nesissary in these lines?

Comment (4e554c4c) [inline src/elf_sections.rs]: I would prefer `core::mem::size_of<ElfSection32|64()` to comparing with `40` and `60` if they are equivilant.

Comment (ahmedcharles) [inline src/elf_sections.rs]: I have more changes I'd like to get in that would clean this one up more, but yes, you have to cast twice, once to get a raw pointer of the same type and again to change the type.

Comment (ahmedcharles) [inline src/elf_sections.rs]: That isn't possible because `match` doesn't accept `size_of` in a pattern.

**Comment by ahmedcharles** [comment]
I updated this to be closer to reasonable. Input welcome.

**Comment by robert-w-gries** [comment]
This PR has been great in my kernel that supports both i686 and x86_64 targets.  Any update on whether this will be merged?

**Comment by phil-opp** [comment]
I'm sorry, I didn't have the time for maintaining this crate over the past months. I'm happy to add maintainers to this project, if someone is willing, to avoid such delays in the future.

**Comment by phil-opp** [comment]
Thanks a lot for this PR, @ahmedcharles! From a quick look, it looks good. I don't use 32bit ELFs personally, so I don't have a good way of testing this feature. I don't have the time to do an in-detail review either. Nevertheless, I'm going to merge this PR since it seems to work, according to the comments in this thread.

I also decided to give you push access to this repo. Feel free to push and merge any non breaking changes at your own discretion.

**Comment by ahmedcharles** [comment]
Thanks.

**Pull request merged**